src/audio/dai-zephyr.c

```diff
@@ -848,12 +848,38 @@ static int dai_set_dma_buffer(struct dai_data *dd, struct comp_dev *dev,
 
 	/* calculate DMA buffer size */
 	period_count = dd->dma->plat_data.period_count;
+
+#if CONFIG_IPC_MAJOR_4
+	/* Use ibs/obs from copier gtw config to calculate DMA buffer length in ms.
+	 * This is used to calculate correct period count value.
+	 * It is especially important for FW aggregation use case, when Copier
+	 * receives whole DMA buffer size in GTW config but should be able to
+	 * allocate multiple DMA's implicitly.
+	 * E.g. if dma_buffer_size received in Copier GTW config is 7680 bytes and FW
+	 * should aggregate two streams 1ch/24_32bit/48kHz and 3ch/24_32bit/48kHz into one
+	 * 4ch/24_32bit/48kHz stream, then FW will have to allocate two DMA's with sizes
+	 * 1920 and 5760 bytes respectively and same 10 ms length (period_count) for both
+	 * to avoid glitches.
```

Comment:
this explanation is very very hard to follow. For playback we typically send the same data to different links. For capture, we typically have a symmetry between the number of channels on different links.

The explanation with 1 and 3 channels does not describe which scenario you are trying to address, and whether this is an actual use case in a product or a test configuration.

In addition, the aggregation is at the link level, and one would hope that we don't have a 10ms buffer?  There's clearly a problem if we again have to rely on more buffers to get this solution to work properly...

Reply:
@plbossart there is no need for any additional buffers. I'm not going to change FW aggregation mechanism.
I found a bug in Windows production scenario (actual use case). When trying to aggregate two streams with different number of channels e.g. 1ch and 3ch (mono and 2.1) and 10ms DMA buffer requested from driver multiple glitches and first stream delay were observed. This change fixes that special case.